packages/core/data-transfer/src/engine/diagnostic.ts

```diff
@@ -0,0 +1,119 @@
+import { EventEmitter } from 'events';
+
+export interface IDiagnosticReporterOptions {
+  stackSize?: number;
+}
+
+export type GenericDiagnostic<K extends DiagnosticKind, T = unknown> = {
+  kind: K;
+  details: {
+    message: string;
+    createdAt: Date;
+  } & T;
+};
+
+export type DiagnosticKind = 'error' | 'warning' | 'info';
+
+export type DiagnosticListener<T extends DiagnosticKind = DiagnosticKind> = (
+  diagnostic: { kind: T } & Diagnostic extends infer U ? U : 'foo'
+) => void | Promise<void>;
+
+export type DiagnosticEvent = 'diagnostic' | `diagnostic.${DiagnosticKind}`;
+
+export type GetEventListener<E extends DiagnosticEvent> = E extends 'diagnostic'
+  ? DiagnosticListener
+  : E extends `diagnostic.${infer K}`
+  ? K extends DiagnosticKind
+    ? DiagnosticListener<K>
+    : never
+  : never;
+
+export type Diagnostic = ErrorDiagnostic | WarningDiagnostic | InfoDiagnostic;
+
+export type ErrorDiagnosticSeverity = 'fatal' | 'error' | 'silly';
+
+export type ErrorDiagnostic = GenericDiagnostic<
+  'error',
+  {
+    name: string;
+    severity: ErrorDiagnosticSeverity;
+    error: Error;
+  }
+>;
+
+export type WarningDiagnostic = GenericDiagnostic<
+  'warning',
+  {
+    origin?: string;
+  }
+>;
+
+export type InfoDiagnostic<T = unknown> = GenericDiagnostic<
+  'info',
+  {
+    params?: T;
+  }
+>;
+
+export interface IDiagnosticReporter {
+  stack: {
+    readonly size: number;
+    readonly items: Diagnostic[];
+  };
+
+  report(diagnostic: Diagnostic): IDiagnosticReporter;
+  onDiagnostic(listener: DiagnosticListener): IDiagnosticReporter;
+  on<T extends DiagnosticKind>(kind: T, listener: DiagnosticListener<T>): IDiagnosticReporter;
+}
+
+const createDiagnosticReporter = (
+  options: IDiagnosticReporterOptions = {}
+): IDiagnosticReporter => {
+  const { stackSize = -1 } = options;
+
+  const emitter = new EventEmitter();
+  const stack: Diagnostic[] = [];
+
+  const addListener = <T extends DiagnosticEvent>(event: T, listener: GetEventListener<T>) => {
+    emitter.on(event, listener);
+  };
+
+  return {
+    stack: {
+      get size() {
+        return stack.length;
+      },
+
+      get items() {
+        return stack;
+      },
+    },
+
+    report(diagnostic: Diagnostic) {
+      emitter.emit('diagnostic', diagnostic);
+      emitter.emit(`diagnostic.${diagnostic.kind}`, diagnostic);
```

Comment:
It feels weird to emit the same event twice, but at the moment I can't think of an alternative